File: internal/codec/images.go
// Package codec provides utilities for converting between API formats.
package codec

import (
	"context"
	"encoding/base64"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/tjfontaine/polyglot-llm-gateway/internal/domain"
)

// ImageFetcher handles fetching remote images and converting them to base64.
type ImageFetcher struct {
	client  *http.Client
	maxSize int64 // Maximum allowed image size in bytes
}

// ImageFetcherOption configures the image fetcher.
type ImageFetcherOption func(*ImageFetcher)

// WithHTTPClient sets a custom HTTP client for the fetcher.
func WithImageHTTPClient(client *http.Client) ImageFetcherOption {
	return func(f *ImageFetcher) {
		f.client = client
	}
}

// WithMaxSize sets the maximum allowed image size.
func WithMaxSize(maxSize int64) ImageFetcherOption {
	return func(f *ImageFetcher) {
		f.maxSize = maxSize
	}
}

// NewImageFetcher creates a new image fetcher.
func NewImageFetcher(opts ...ImageFetcherOption) *ImageFetcher {
	f := &ImageFetcher{
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
		maxSize: 20 * 1024 * 1024, // 20MB default max
	}
	for _, opt := range opts {
		opt(f)
	}
	return f
}

// FetchAndConvert fetches an image from a URL and converts it to base64 format.
// Returns an ImageSource suitable for Anthropic's API.
func (f *ImageFetcher) FetchAndConvert(ctx context.Context, url string) (*domain.ImageSource, error) {
	// Check if it's already a data URL (base64 encoded)
	if strings.HasPrefix(url, "data:") {
		return f.parseDataURL(url)
	}

	// Validate URL scheme
	if !strings.HasPrefix(url, "http://") && !strings.HasPrefix(url, "https://") {
		return nil, fmt.Errorf("unsupported URL scheme: must be http:// or https://")
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := f.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch image: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch image: status %d", resp.StatusCode)
	}

	// Check content length if available
	if resp.ContentLength > f.maxSize {
		return nil, fmt.Errorf("image too large: %d bytes (max %d)", resp.ContentLength, f.maxSize)
	}

	// Determine media type from Content-Type header
	mediaType := resp.Header.Get("Content-Type")
	if mediaType == "" {
		// Try to infer from URL
		mediaType = inferMediaType(url)
	}

	// Validate media type
	if !isSupportedMediaType(mediaType) {
		return nil, fmt.Errorf("unsupported media type: %s", mediaType)
	}

	// Read the image data with size limit
	limitReader := io.LimitReader(resp.Body, f.maxSize+1)
	data, err := io.ReadAll(limitReader)
	if err != nil {
		return nil, fmt.Errorf("failed to read image: %w", err)
	}

	if int64(len(data)) > f.maxSize {
		return nil, fmt.Errorf("image too large: exceeds %d bytes", f.maxSize)
	}

	// Encode to base64
	encoded := base64.StdEncoding.EncodeToString(data)

	return &domain.ImageSource{
		Type:      "base64",
		MediaType: normalizeMediaType(mediaType),
		Data:      encoded,
	}, nil
}

// parseDataURL parses a data URL and extracts the base64 content.
func (f *ImageFetcher) parseDataURL(url string) (*domain.ImageSource, error) {
	// Format: data:image/jpeg;base64,/9j/4AAQSkZ...
	if !strings.HasPrefix(url, "data:") {
		return nil, fmt.Errorf("not a data URL")
	}

	// Remove "data:" prefix
	content := url[5:]

	// Find the comma separator
	commaIdx := strings.Index(content, ",")
	if commaIdx == -1 {
		return nil, fmt.Errorf("invalid data URL: missing comma separator")
	}

	metadata := content[:commaIdx]
	data := content[commaIdx+1:]

	// Parse metadata
	parts := strings.Split(metadata, ";")
	if len(parts) == 0 {
		return nil, fmt.Errorf("invalid data URL: missing media type")
	}

	mediaType := parts[0]
	if !isSupportedMediaType(mediaType) {
		return nil, fmt.Errorf("unsupported media type: %s", mediaType)
	}

	// Check for base64 encoding
	isBase64 := false
	for _, part := range parts[1:] {
		if part == "base64" {
			isBase64 = true
			break
		}
	}

	if !isBase64 {
		return nil, fmt.Errorf("data URL must be base64 encoded")
	}

	return &domain.ImageSource{
		Type:      "base64",
		MediaType: normalizeMediaType(mediaType),
		Data:      data,
	}, nil
}

// inferMediaType attempts to infer the media type from a URL.
func inferMediaType(url string) string {
	urlLower := strings.ToLower(url)

	switch {
	case strings.HasSuffix(urlLower, ".jpg") || strings.HasSuffix(urlLower, ".jpeg"):
		return "image/jpeg"
	case strings.HasSuffix(urlLower, ".png"):
		return "image/png"
	case strings.HasSuffix(urlLower, ".gif"):
		return "image/gif"
	case strings.HasSuffix(urlLower, ".webp"):
		return "image/webp"
	default:
		return "image/jpeg" // Default assumption
	}
}

// isSupportedMediaType checks if the media type is supported by Anthropic.
func isSupportedMediaType(mediaType string) bool {
	// Normalize by taking only the main type (ignore parameters like charset)
	mainType := strings.Split(mediaType, ";")[0]
	mainType = strings.TrimSpace(strings.ToLower(mainType))

	switch mainType {
	case "image/jpeg", "image/jpg", "image/png", "image/gif", "image/webp":
		return true
	default:
		return false
	}
}

// normalizeMediaType normalizes the media type to a standard format.
func normalizeMediaType(mediaType string) string {
	mainType := strings.Split(mediaType, ";")[0]
	mainType = strings.TrimSpace(strings.ToLower(mainType))

	// Normalize image/jpg to image/jpeg
	if mainType == "image/jpg" {
		return "image/jpeg"
	}
	return mainType
}

// ConvertContentPart converts an image_url content part to base64 image for Anthropic.
// If the content part is not an image_url or conversion fails, returns the original.
func (f *ImageFetcher) ConvertContentPart(ctx context.Context, part *domain.ContentPart) (*domain.ContentPart, error) {
	if part.Type != domain.ContentTypeImageURL || part.ImageURL == nil {
		return part, nil
	}

	source, err := f.FetchAndConvert(ctx, part.ImageURL.URL)
	if err != nil {
		return nil, fmt.Errorf("failed to convert image URL: %w", err)
	}

	return &domain.ContentPart{
		Type:   domain.ContentTypeImage,
		Source: source,
	}, nil
}

File: ts/packages/gateway-core/src/codecs/images.ts
/**
 * Image fetcher for converting remote images to base64.
 *
 * Required for OpenAI→Anthropic translation where image URLs
 * need to be converted to base64 format.
 *
 * @module codec/images
 */

// ============================================================================
// Types
// ============================================================================

/**
 * Base64 image source for Anthropic API.
 * Named Base64ImageSource to avoid conflict with domain ImageSource.
 */
export interface Base64ImageSource {
    /** Source type. */
    type: 'base64';

    /** Media type (e.g., image/jpeg). */
    mediaType: string;

    /** Base64 encoded image data. */
    data: string;
}

/**
 * Image fetcher options.
 */
export interface ImageFetcherOptions {
    /** Custom fetch function. */
    fetch?: typeof fetch | undefined;

    /** Maximum allowed image size in bytes (default: 20MB). */
    maxSize?: number | undefined;

    /** Request timeout in milliseconds (default: 10000). */
    timeoutMs?: number | undefined;
}

// ============================================================================
// Image Fetcher
// ============================================================================

/**
 * Default max image size (20MB).
 */
const DEFAULT_MAX_SIZE = 20 * 1024 * 1024;

/**
 * Default request timeout (10 seconds).
 */
const DEFAULT_TIMEOUT_MS = 10000;

/**
 * Supported media types for images.
 */
const SUPPORTED_MEDIA_TYPES = new Set([
    'image/jpeg',
    'image/jpg',
    'image/png',
    'image/gif',
    'image/webp',
]);

/**
 * Fetches and converts images to base64 format.
 */
export class ImageFetcher {
    private readonly fetchFn: typeof fetch;
    private readonly maxSize: number;
    private readonly timeoutMs: number;

    constructor(options: ImageFetcherOptions = {}) {
        this.fetchFn = options.fetch ?? globalThis.fetch.bind(globalThis);
        this.maxSize = options.maxSize ?? DEFAULT_MAX_SIZE;
        this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    }

    /**
     * Fetches an image from URL and converts to base64.
     */
    async fetchAndConvert(url: string): Promise<Base64ImageSource> {
        // Handle data URLs
        if (url.startsWith('data:')) {
            return this.parseDataUrl(url);
        }

        // Validate URL scheme
        if (!url.startsWith('http://') && !url.startsWith('https://')) {
            throw new Error('Unsupported URL scheme: must be http:// or https://');
        }

        // Fetch with timeout
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);

        try {
            const response = await this.fetchFn(url, {
                signal: controller.signal,
            });

            if (!response.ok) {
                throw new Error(`Failed to fetch image: status ${response.status}`);
            }

            // Check content length if available
            const contentLength = response.headers.get('content-length');
            if (contentLength && parseInt(contentLength, 10) > this.maxSize) {
                throw new Error(
                    `Image too large: ${contentLength} bytes (max ${this.maxSize})`,
                );
            }

            // Determine media type
            let mediaType = response.headers.get('content-type') ?? '';
            if (!mediaType) {
                mediaType = this.inferMediaType(url);
            }

            // Validate media type
            const normalizedType = this.normalizeMediaType(mediaType);
            if (!SUPPORTED_MEDIA_TYPES.has(normalizedType)) {
                throw new Error(`Unsupported media type: ${mediaType}`);
            }

            // Read and check size
            const buffer = await response.arrayBuffer();
            if (buffer.byteLength > this.maxSize) {
                throw new Error(`Image too large: exceeds ${this.maxSize} bytes`);
            }

            // Encode to base64
            const encoded = this.arrayBufferToBase64(buffer);

            return {
                type: 'base64',
                mediaType: normalizedType,
                data: encoded,
            };
        } finally {
            clearTimeout(timeoutId);
        }
    }

    /**
     * Parses a data URL to extract base64 content.
     */
    private parseDataUrl(url: string): Base64ImageSource {
        // Format: data:image/jpeg;base64,/9j/4AAQSkZ...
        if (!url.startsWith('data:')) {
            throw new Error('Not a data URL');
        }

        const content = url.slice(5);
        const commaIdx = content.indexOf(',');
        if (commaIdx === -1) {
            throw new Error('Invalid data URL: missing comma separator');
        }

        const metadata = content.slice(0, commaIdx);
        const data = content.slice(commaIdx + 1);

        const parts = metadata.split(';');
        if (parts.length === 0) {
            throw new Error('Invalid data URL: missing media type');
        }

        const mediaType = this.normalizeMediaType(parts[0]!);
        if (!SUPPORTED_MEDIA_TYPES.has(mediaType)) {
            throw new Error(`Unsupported media type: ${parts[0]}`);
        }

        // Check for base64 encoding
        if (!parts.includes('base64')) {
            throw new Error('Data URL must be base64 encoded');
        }

        return {
            type: 'base64',
            mediaType,
            data,
        };
    }

    /**
     * Infers media type from URL extension.
     */
    private inferMediaType(url: string): string {
        const lower = url.toLowerCase();

        if (lower.endsWith('.jpg') || lower.endsWith('.jpeg')) {
            return 'image/jpeg';
        }
        if (lower.endsWith('.png')) {
            return 'image/png';
        }
        if (lower.endsWith('.gif')) {
            return 'image/gif';
        }
        if (lower.endsWith('.webp')) {
            return 'image/webp';
        }

        return 'image/jpeg'; // Default assumption
    }

    /**
     * Normalizes media type.
     */
    private normalizeMediaType(mediaType: string): string {
        const main = mediaType.split(';')[0]?.trim().toLowerCase() ?? '';

        // Normalize image/jpg to image/jpeg
        if (main === 'image/jpg') {
            return 'image/jpeg';
        }

        return main;
    }

    /**
     * Converts ArrayBuffer to base64 string.
     */
    private arrayBufferToBase64(buffer: ArrayBuffer): string {
        const bytes = new Uint8Array(buffer);

        // For environments with Buffer (Node.js)
        if (typeof Buffer !== 'undefined') {
            return Buffer.from(bytes).toString('base64');
        }

        // For browser/Cloudflare Workers
        let binary = '';
        for (let i = 0; i < bytes.byteLength; i++) {
            binary += String.fromCharCode(bytes[i]!);
        }
        return btoa(binary);
    }
}

/**
 * Creates an image fetcher with default options.
 */
export function createImageFetcher(options?: ImageFetcherOptions): ImageFetcher {
    return new ImageFetcher(options);
}
